src/simulator_core/entities/assets/demand_cluster.py

```diff
@@ -29,50 +54,159 @@ def __init__(
+        - A heat exchanger to set the power consumption of the demand
+        - An intermediate junction to link both components.
+        :param pandapipesNet pandapipes_net: pandapipes network object
+        """
+        if not self._initialized:
+            # Create intermediate junction
+            self._intermediate_junction = Junction(
+                pandapipes_net=self.pandapipes_net,
+                pn_bar=self.pressure_input,
+                tfluid_k=self.temperature_supply,
+                height_m=DEFAULT_NODE_HEIGHT,
+                name=f"intermediate_junction_{self.name}",
+            )
+            # Create the control valve
+            self._flow_control = ControlValve(
+                pandapipes_net=self.pandapipes_net,
+                controlled_mdot_kg_per_s=self.mass_flowrate,
+                diameter_m=self.internal_diameter,
+                control_active=True,
+                in_service=True,
+                name=f"flow_control_{self.name}",
+            )
+            self._flow_control.from_junction = self._from_junction
+            self._flow_control.to_junction = self._intermediate_junction
+            self._flow_control.create()
+
+            # Create the heat exchanger
+            self._heat_exchanger = HeatExchanger(
+                pandapipes_net=self.pandapipes_net,
+                diameter_m=self.internal_diameter,
+                heat_flux_w=self.thermal_power_allocation,
+                name=f"heat_exchanger_{self.name}",
+                in_service=True,
+            )
+            self._heat_exchanger.from_junction = self._intermediate_junction
+            self._heat_exchanger.to_junction = self._to_junction
+            self._heat_exchanger.create()
+
+            self._initialized = True
+
+    def set_setpoints(self, setpoints: Dict) -> None:
         """Placeholder to set the setpoints of an asset prior to a simulation.
 
         :param Dict setpoints: The setpoints that should be set for the asset.
             The keys of the dictionary are the names of the setpoints and the values are the values
         """
-        pass
-
-    def get_setpoints(self, **kwargs) -> Dict:
+        self.thermal_power_allocation = setpoints[PROPERTY_HEAT_DEMAND]
+        self.temperature_return_target = setpoints[PROPERTY_TEMPERATURE_RETURN]
+
+        if self.simulation_performed():
+            self.temperature_supply = self.pandapipes_net.res_flow_control['t_from_k'][
+                self._flow_control.index]
+            self.temperature_return = self.pandapipes_net.res_heat_exchanger['t_to_k'][
+                self._heat_exchanger.index]
+            self.mass_flowrate = self.pandapipes_net.res_heat_exchanger['mdot_from_kg_per_s'][
+                self._heat_exchanger.index]
+
+        # adjust flowrate or power to meet the return temperature
+        if self.pandapipes_net.flow_control.control_active[self._flow_control.index] is True:
```

Comment:
Can you make two seperate methods for this? Like _set_mass_flow_pump_enabled and _set_mass_flow_pump_disabled?

Reply:
i create _set_demand_control